# Capstone Project

## EDA for Car Price Prediction Model

---

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib as plt
import warnings
import warnings
warnings.filterwarnings('ignore')

In [2]:
df = pd.read_json('scout_car.json', lines=True)

In [3]:
df.sample(20).T

1613   \
url                            https://www.autoscout24.com//offers/audi-a1-sp...   
make_model                                                               Audi A1   
short_description                     Sportback 1.4 TFSI FSE KLIMA PDC SHZ EURO6   
body_type                                                                Compact   
price                                                                      22440   
vat                                                               VAT deductible   
km                                                                      4,200 km   
registration                                                             06/2018   
prev_owner                                                      1 previous owner   
kW                                                                           NaN   
hp                                                                         92 kW   
Type                                              [, Employee's car, , Super 95]   
Previous Owners                                                            \n1\n   
Next Inspection                           [\n06/2021\n, \n123 g CO2/km (comb)\n]   
Inspection new                                                               NaN   
Warranty                                                    [\n, \n, \nEuro 6\n]   
Full Service                                                                 NaN   
Non-smoking Vehicle                                                          NaN   
null                                                                          []   
Make                                                                    \nAudi\n   
Model                                                               [\n, A1, \n]   
Offer Number                                                       [\nAKT5268\n]   
First Registration                                                [\n, 2018, \n]   
Body Color                                                      [\n, Silver, \n]   
Paint Type                                                        [\nMetallic\n]   
Body Color Original                                          [\nflorettsilber\n]   
Upholstery                                                    [\nCloth, Black\n]   
Body                                                           [\n, Compact, \n]   
Nr. of Doors                                                             [\n5\n]   
Nr. of Seats                                                             [\n5\n]   
Model Code                                                        [\n0588/BCW\n]   
Gearing Type                                                 [\n, Automatic, \n]   
Displacement                                                      [\n1,395 cc\n]   
Cylinders                                                                [\n4\n]   
Weight                                                                       NaN   
Drive chain                                                          [\nfront\n]   
Fuel                                                          [\n, Super 95, \n]   
Consumption                    [[5.2 l/100 km (comb)], [6.7 l/100 km (city)],...   
CO2 Emission                                           [\n123 g CO2/km (comb)\n]   
Emission Class                                                      [\nEuro 6\n]   
\nComfort & Convenience\n      [Air conditioning, Armrest, Automatic climate ...   
\nEntertainment & Media\n      [Bluetooth, CD player, Hands-free equipment, M...   
\nExtras\n                     [Alloy wheels, Catalytic Converter, Sport susp...   
\nSafety & Security\n          [ABS, Central door lock, Daytime running light...   
description                    [\nFahrzeug-Nr.: AKT5268., Media:, Navigations...   
Emission Label                                                   [\n4 (Green)\n]   
Gears                                                                        NaN   
Country version                            

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15919 entries, 0 to 15918
Data columns (total 54 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   url                            15919 non-null  object 
 1   make_model                     15919 non-null  object 
 2   short_description              15873 non-null  object 
 3   body_type                      15859 non-null  object 
 4   price                          15919 non-null  int64  
 5   vat                            11406 non-null  object 
 6   km                             15919 non-null  object 
 7   registration                   15919 non-null  object 
 8   prev_owner                     9091 non-null   object 
 9   kW                             0 non-null      float64
 10  hp                             15919 non-null  object 
 11  Type                           15917 non-null  object 
 12  Previous Owners                9279 non-null  

In [5]:
df.short_description.describe()

count                                 15873
unique                                10001
top       SPB 1.6 TDI 116 CV S tronic Sport
freq                                     64
Name: short_description, dtype: object

In [6]:
df.short_description.unique()

array(['Sportback 1.4 TDI S-tronic Xenon Navi Klima', '1.8 TFSI sport',
       'Sportback 1.6 TDI S tronic Einparkhilfe plus+music', ...,
       'ELYSEE ENERGY dCi 160 EDC',
       'INITIALE Paris TCe 225 EDC GPF ACC EU6',
       'TCe 225 EDC GPF LIM Deluxe Pano,RFK'], dtype=object)

In [7]:
df.body_type.describe()

count      15859
unique         9
top       Sedans
freq        7903
Name: body_type, dtype: object

In [8]:
df.body_type.unique()

array(['Sedans', 'Station wagon', 'Compact', 'Other', 'Coupe', 'Van',
       'Off-Road', 'Convertible', None, 'Transporter'], dtype=object)

In [9]:
df.body_type.isna().sum()

60

In [10]:
df.price.describe()

count    15919.000000
mean     18019.896727
std       7386.169409
min         13.000000
25%      12850.000000
50%      16900.000000
75%      21900.000000
max      74600.000000
Name: price, dtype: float64

In [11]:
df[df.price<5000].T

6066  \
url                            https://www.autoscout24.com//offers/opel-astra...   
make_model                                                            Opel Astra   
short_description                 1.6 D (CDTI) Start/Stop Sports Tourer Business   
body_type                                                          Station wagon   
price                                                                        255   
vat                                                                         None   
km                                                                      5,563 km   
registration                                                             06/2018   
prev_owner                                                                  None   
kW                                                                           NaN   
hp                                                                        100 kW   
Type                                     [, Used, , Diesel (Particulate Filter)]   
Previous Owners                                                              NaN   
Next Inspection                                                              NaN   
Inspection new                                                               NaN   
Warranty                                                                     NaN   
Full Service                                                                 NaN   
Non-smoking Vehicle                                                          NaN   
null                                                                          []   
Make                                                                    \nOpel\n   
Model                                                            [\n, Astra, \n]   
Offer Number                                                                 NaN   
First Registration                                                [\n, 2018, \n]   
Body Color                                                        [\n, Gold, \n]   
Paint Type                                                                   NaN   
Body Color Original                                                          NaN   
Upholstery                                                                   NaN   
Body                                                     [\n, Station wagon, \n]   
Nr. of Doors                                                             [\n5\n]   
Nr. of Seats                                                             [\n5\n]   
Model Code                                                                   NaN   
Gearing Type                                                    [\n, Manual, \n]   
Displacement                                                      [\n1,598 cc\n]   
Cylinders                                                                [\n4\n]   
Weight                                                            [\n1,403 kg\n]   
Drive chain                                                                  NaN   
Fuel                                       [\n, Diesel (Particulate Filter), \n]   
Consumption                    [[3.8 l/100 km (comb)], [4.5 l/100 km (city)],...   
CO2 Emission                                           [\n101 g CO2/km (comb)\n]   
Emission Class                                                      [\nEuro 6\n]   
\nComfort & Convenience\n      [Air conditioning, Cruise control, Electrical ...   
\nEntertainment & Media\n      [Bluetooth, Hands-free equipment, On-board com...   
\nExtras\n                                                           [Roof rack]   
\nSafety & Security\n          [ABS, Central door lock with remote control, D...   
description                                                                   []   
Emission Label                                                               NaN   
Gears                                                                    [\n6\n]   
Country version                             

In [12]:
df.sample(3).T

14514  \
url                            https://www.autoscout24.com//offers/renault-cl...   
make_model                                                          Renault Clio   
short_description                                   IV SOCIETE DCI 75 ENERGY AIR   
body_type                                                            Transporter   
price                                                                      10797   
vat                                                                         None   
km                                                                          1 km   
registration                                                             03/2018   
prev_owner                                                                  None   
kW                                                                           NaN   
hp                                                                         55 kW   
Type                                                          [, Used, , Diesel]   
Previous Owners                                                              NaN   
Next Inspection                                                              NaN   
Inspection new                                                               NaN   
Warranty                                                            \n9 months\n   
Full Service                                                                 NaN   
Non-smoking Vehicle                                                          NaN   
null                                                                          []   
Make                                                                 \nRenault\n   
Model                                                             [\n, Clio, \n]   
Offer Number                                                        [\n163465\n]   
First Registration                                                [\n, 2018, \n]   
Body Color                                                       [\n, White, \n]   
Paint Type                                                                   NaN   
Body Color Original                                                  [\nblanc\n]   
Upholstery                                                                   NaN   
Body                                                       [\n, Transporter, \n]   
Nr. of Doors                                                             [\n5\n]   
Nr. of Seats                                                             [\n2\n]   
Model Code                                                                   NaN   
Gearing Type                                                    [\n, Manual, \n]   
Displacement                                                      [\n1,461 cc\n]   
Cylinders                                                                    NaN   
Weight                                                                       NaN   
Drive chain                                                                  NaN   
Fuel                                                            [\n, Diesel, \n]   
Consumption                                                                  NaN   
CO2 Emission                                                        [[], [], []]   
Emission Class                                                               NaN   
\nComfort & Convenience\n                                       [Cruise control]   
\nEntertainment & Media\n                                                    NaN   
\nExtras\n                                                                   NaN   
\nSafety & Security\n          [Driver-side airbag, Passenger-side airbag, Si...   
description                       [\nStarterre, Centre Automobile Multimarque\n]   
Emission Label                                                               NaN   
Gears                                                                        NaN   
Country version                            

In [13]:
df.vat.describe()

count              11406
unique                 2
top       VAT deductible
freq               10980
Name: vat, dtype: object

In [14]:
df.vat.unique()

array(['VAT deductible', 'Price negotiable', None], dtype=object)

In [15]:
df.registration.describe()

count     15919
unique       48
top         -/-
freq       1597
Name: registration, dtype: object

In [16]:
df.registration.unique()

array(['01/2016', '03/2017', '02/2016', '08/2016', '05/2016', '03/2016',
       '06/2017', '05/2018', '09/2016', '06/2016', '10/2016', '04/2016',
       '06/2018', '11/2017', '07/2016', '12/2016', '09/2018', '04/2018',
       '07/2017', '02/2017', '11/2016', '01/2018', '05/2017', '04/2017',
       '07/2018', '01/2017', '02/2018', '03/2018', '-/-', '08/2017',
       '10/2017', '08/2018', '09/2017', '12/2017', '04/2019', '12/2018',
       '03/2019', '06/2019', '05/2019', '02/2019', '01/2019', '10/2018',
       '11/2018', '09/2019', '07/2019', '08/2019', '11/2019', '12/2019'],
      dtype=object)

In [17]:
df.prev_owner.describe()

count                 9091
unique                   4
top       1 previous owner
freq                  8294
Name: prev_owner, dtype: object

In [18]:
df.prev_owner.unique()

array(['2 previous owners', None, '1 previous owner', '3 previous owners',
       '4 previous owners'], dtype=object)

In [19]:
df.prev_owner.value_counts()

1 previous owner     8294
2 previous owners     778
3 previous owners      17
4 previous owners       2
Name: prev_owner, dtype: int64

In [20]:
df.prev_owner.isnull().sum()

6828

In [21]:
df['Previous Owners'].describe()

count      9279
unique      102
top       \n1\n
freq       8101
Name: Previous Owners, dtype: object

In [22]:
#df['Previous Owners'].value_counts(dropna=False)

In [23]:
df.Type.describe()

count                                       15917
unique                                        169
top       [, Used, , Diesel (Particulate Filter)]
freq                                         3475
Name: Type, dtype: object

In [24]:
df['Next Inspection'].describe()

count            3535
unique           1399
top       \n04/2022\n
freq               62
Name: Next Inspection, dtype: object

In [25]:
df['Inspection new'].describe()

count                      3932
unique                      202
top       [\nYes\n, \nEuro 6\n]
freq                        523
Name: Inspection new, dtype: object

In [26]:
df.Warranty.describe()

count                    10499
unique                     515
top       [\n, \n, \nEuro 6\n]
freq                      1868
Name: Warranty, dtype: object

In [27]:
df.Warranty.str.contains("CO", case = False).value_counts()

False    3036
Name: Warranty, dtype: int64

In [28]:
df.Warranty.isna().sum()

5420

In [29]:
df.Warranty.dtype.type

numpy.object_

In [30]:
df['Full Service'].describe()

count                        8215
unique                        122
top       [\n, \n, \n4 (Green)\n]
freq                         2235
Name: Full Service, dtype: object

In [31]:
df['Non-smoking Vehicle'].describe()

count         7177
unique          94
top       [\n, \n]
freq          3647
Name: Non-smoking Vehicle, dtype: object

In [32]:
df['Non-smoking Vehicle'].isnull().sum()

8742

In [33]:
df['Make'].describe()

count        15919
unique           3
top       \nOpel\n
freq          7343
Name: Make, dtype: object

In [34]:
df['Make'].unique()

array(['\nAudi\n', '\nOpel\n', '\nRenault\n'], dtype=object)

In [35]:
df['Model'].describe()

count            15919
unique               9
top       [\n, A3, \n]
freq              3097
Name: Model, dtype: object

In [36]:
df['First Registration'].describe()

count              14322
unique                 4
top       [\n, 2018, \n]
freq                4522
Name: First Registration, dtype: object

In [37]:
df['First Registration'].value_counts()

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas\_libs\hashtable_class_helper.pxi", line 4588, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


[\n, 2018, \n]    4522
[\n, 2016, \n]    3674
[\n, 2017, \n]    3273
[\n, 2019, \n]    2853
Name: First Registration, dtype: int64

In [38]:
df.registration.isin(['2018']).value_counts()

False    15919
Name: registration, dtype: int64

In [39]:
df.registration.value_counts()

-/-        1597
03/2018     695
02/2019     585
05/2018     572
03/2019     543
01/2019     541
04/2018     541
02/2018     539
03/2016     536
04/2016     532
06/2018     532
01/2018     511
04/2019     506
02/2016     472
03/2017     471
05/2016     459
06/2016     452
05/2019     440
06/2017     409
05/2017     404
07/2018     396
04/2017     380
01/2016     376
02/2017     368
01/2017     306
08/2018     285
06/2019     224
07/2017     215
11/2017     180
07/2016     176
10/2016     160
10/2017     154
09/2017     149
11/2016     142
09/2016     141
09/2018     141
12/2016     134
12/2017     123
08/2017     114
11/2018     110
12/2018     103
10/2018      97
08/2016      94
07/2019       6
09/2019       5
08/2019       1
11/2019       1
12/2019       1
Name: registration, dtype: int64

In [40]:
df.registration.isnull().sum()

0

In [41]:
df['Body Color'].describe()

count               15322
unique                 14
top       [\n, Black, \n]
freq                 3745
Name: Body Color, dtype: object

In [42]:
df['Body Color'].value_counts()

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas\_libs\hashtable_class_helper.pxi", line 4588, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


[\n, Black, \n]     3745
[\n, Grey, \n]      3505
[\n, White, \n]     3406
[\n, Silver, \n]    1647
[\n, Blue, \n]      1431
[\n, Red, \n]        957
[\n, Brown, \n]      289
[\n, Green, \n]      154
[\n, Beige, \n]      108
[\n, Yellow, \n]      51
[\n, Violet, \n]      18
[\n, Bronze, \n]       6
[\n, Orange, \n]       3
[\n, Gold, \n]         2
Name: Body Color, dtype: int64

In [43]:
df['Body Color Original'].describe()

count                  12160
unique                  1927
top       [\nOnyx Schwarz\n]
freq                     338
Name: Body Color Original, dtype: object

In [44]:
df['Paint Type'].describe()

count              10147
unique                 3
top       [\nMetallic\n]
freq                9794
Name: Paint Type, dtype: object

In [45]:
df['Paint Type'].isnull().sum()

5772

In [46]:
df['Upholstery'].value_counts()

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas\_libs\hashtable_class_helper.pxi", line 4588, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


[\nCloth, Black\n]           5821
[\nPart leather, Black\n]    1121
[\nCloth\n]                  1005
[\nCloth, Grey\n]             891
[\nCloth, Other\n]            639
[\nFull leather, Black\n]     575
[\nBlack\n]                   491
[\nGrey\n]                    273
[\nOther, Other\n]            182
[\nPart leather\n]            140
[\nFull leather\n]            139
[\nFull leather, Brown\n]     116
[\nPart leather, Grey\n]      116
[\nOther, Black\n]            110
[\nFull leather, Other\n]      72
[\nFull leather, Grey\n]       67
[\nPart leather, Other\n]      65
[\nOther\n]                    56
[\nPart leather, Brown\n]      50
[\nalcantara, Black\n]         47
[\nVelour, Black\n]            36
[\nFull leather, Beige\n]      36
[\nCloth, Brown\n]             28
[\nVelour\n]                   16
[\nOther, Grey\n]              15
[\nCloth, Beige\n]             13
[\nBrown\n]                    12
[\nCloth, Blue\n]              12
[\nCloth, White\n]              8
[\nVelour, Gre

In [47]:
df['Nr. of Doors'].describe()

count       15707
unique          6
top       [\n5\n]
freq        11575
Name: Nr. of Doors, dtype: object

In [48]:
df['Nr. of Doors'].value_counts()

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas\_libs\hashtable_class_helper.pxi", line 4588, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


[\n5\n]    11575
[\n4\n]     3079
[\n3\n]      832
[\n2\n]      219
[\n7\n]        1
[\n1\n]        1
Name: Nr. of Doors, dtype: int64

In [49]:
df['Nr. of Seats'].value_counts()

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas\_libs\hashtable_class_helper.pxi", line 4588, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


[\n5\n]    13336
[\n4\n]     1125
[\n7\n]      362
[\n2\n]      116
[\n6\n]        2
[\n3\n]        1
Name: Nr. of Seats, dtype: int64

In [50]:
df['Gearing Type'].value_counts()

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas\_libs\hashtable_class_helper.pxi", line 4588, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


[\n, Manual, \n]            8153
[\n, Automatic, \n]         7297
[\n, Semi-automatic, \n]     469
Name: Gearing Type, dtype: int64

In [51]:
df.Gears.describe()

count       11207
unique         10
top       [\n6\n]
freq         5822
Name: Gears, dtype: object

In [52]:
df.Gears.value_counts()

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas\_libs\hashtable_class_helper.pxi", line 4588, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


[\n6\n]     5822
[\n5\n]     3239
[\n7\n]     1908
[\n8\n]      224
[\n9\n]        6
[\n4\n]        2
[\n3\n]        2
[\n1\n]        2
[\n50\n]       1
[\n2\n]        1
Name: Gears, dtype: int64

In [53]:
df.Displacement.describe()

count              15423
unique                77
top       [\n1,598 cc\n]
freq                4761
Name: Displacement, dtype: object

In [54]:
df.Cylinders.describe()

count       10239
unique          7
top       [\n4\n]
freq         8105
Name: Cylinders, dtype: object

In [55]:
df.Cylinders.value_counts()

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas\_libs\hashtable_class_helper.pxi", line 4588, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


[\n4\n]    8105
[\n3\n]    2104
[\n5\n]      22
[\n6\n]       3
[\n2\n]       2
[\n8\n]       2
[\n1\n]       1
Name: Cylinders, dtype: int64

In [56]:
df.Weight.describe()

count               8945
unique               434
top       [\n1,163 kg\n]
freq                 574
Name: Weight, dtype: object

In [57]:
df['Drive chain'].describe()

count            9061
unique              3
top       [\nfront\n]
freq             8886
Name: Drive chain, dtype: object

In [58]:
df['Drive chain'].value_counts()

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas\_libs\hashtable_class_helper.pxi", line 4588, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


[\nfront\n]    8886
[\n4WD\n]       171
[\nrear\n]        4
Name: Drive chain, dtype: int64

In [59]:
df.Fuel.describe()

count                                     15919
unique                                       77
top       [\n, Diesel (Particulate Filter), \n]
freq                                       4315
Name: Fuel, dtype: object

In [60]:
df['Consumption'].describe()

count                                                 14013
unique                                                  881
top       [[3.9 l/100 km (comb)], [4.1 l/100 km (city)],...
freq                                                    304
Name: Consumption, dtype: object

In [61]:
df['Consumption'].iloc[3]

[['3.8 l/100 km (comb)'], ['4.3 l/100 km (city)'], ['3.5 l/100 km (country)']]

In [62]:
df['CO2 Emission'].describe()

count                         14111
unique                          123
top       [\n120 g CO2/km (comb)\n]
freq                            740
Name: CO2 Emission, dtype: object

In [63]:
df['Emission Class'].describe()

count            12898
unique               7
top       [\nEuro 6\n]
freq             10139
Name: Emission Class, dtype: object

In [64]:
df['Emission Class'].value_counts()

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas\_libs\hashtable_class_helper.pxi", line 4588, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


[\nEuro 6\n]          10139
[\nEuro 6d-TEMP\n]     1845
[[], [], []]            607
[\nEuro 6c\n]           127
[\nEuro 5\n]             78
[\nEuro 6d\n]            62
[\nEuro 4\n]             40
Name: Emission Class, dtype: int64

In [65]:
df['\nComfort & Convenience\n'].describe()

count                                                 14999
unique                                                 6198
top       [Air conditioning, Electrical side mirrors, Hi...
freq                                                    216
Name: \nComfort & Convenience\n, dtype: object

In [66]:
df['\nEntertainment & Media\n'].describe()

count                                                 14545
unique                                                  346
top       [Bluetooth, Hands-free equipment, On-board com...
freq                                                   1282
Name: \nEntertainment & Media\n, dtype: object

In [67]:
df['\nExtras\n'].describe()

count              12957
unique               659
top       [Alloy wheels]
freq                3245
Name: \nExtras\n, dtype: object

In [68]:
df['\nSafety & Security\n'].describe()

count                                                 14937
unique                                                 4443
top       [ABS, Central door lock, Daytime running light...
freq                                                    538
Name: \nSafety & Security\n, dtype: object

In [69]:
df.description.describe()

count     15919
unique    12874
top          []
freq        636
Name: description, dtype: object

### Things Needs to Be Done

1. Think about removing url column.
2. **Get rid of the column "make_model"**
3. **short description
4. Body_type : There are 60 nulls values that we have to deal with.
5. Be aware of outliers such as min. in price column.
6. Vat column : Change it to deductible or not deductible
7. For the km column remove the "km" from the rows and change the data type to int or float.
8. **Drop the registration column because there is another registratio colum which only contains years.**
9. **From prev_owner column remove the part of "previous owner" and change data type to integer. There are 6828 null values.Remove the "Previous owners" column.**
10. **Drop the column "kW"**
11. Change the "hp" column name to kW.Change the data type.
12. Clean the "Type" column.
13. **Drop the columns about inspectation.**
14. Come back to the Warranty column
15. **Drop the Full service column.**
16. **Drop the Non_smoking column.**
17. **Drop the null column.**
18. Clean both make and model columns.
19. Come back to the offer number column again.
20. Change the First registration column name as Registration year
21. **Drop the Body color original**
22. Organize Upholstery column
23. **Drop the Body column.**
24. Check the number of doors
25. For the number od seats, clean '\n'
26. **Drop the model code number  column**
27. Clean the outliers for gearing type column
28. For the Cyclinder column clean the outliers.
29. **You might drop Weight column(!)**
30. Make the small groups for Fuel column.
31. We might need the 'Consumption' to be in 3 different column.
32. For the CO2 Emission, get rid of the unnecessary string parts and change the column name.
33. For the Emission Class column be aware of empty brackets which are null valus and get rid of unnecessary inf.
34. Come back to the \nComfort & Convenience\n column.
35. Come back to '\nEntertainment & Media\n' column.
36. Extras column*
37. Safety & Security column*
38. **Drop the description column.**
39. **Drop the emission column.**
40. Gears column*
41. **Drop rest of the columns.**

In [70]:
df.sample(3).T

2573   \
url                            https://www.autoscout24.com//offers/audi-a1-sp...   
make_model                                                               Audi A1   
short_description                Sportback 25 TFSI Aktionsmodell LED PDC LM Klim   
body_type                                                                Compact   
price                                                                      21990   
vat                                                               VAT deductible   
km                                                                          - km   
registration                                                                 -/-   
prev_owner                                                                  None   
kW                                                                           NaN   
hp                                                                         70 kW   
Type                                                         [, New, , Super 95]   
Previous Owners                                                              NaN   
Next Inspection                                                              NaN   
Inspection new                                                               NaN   
Warranty                                                                     NaN   
Full Service                                                                 NaN   
Non-smoking Vehicle                                                          NaN   
null                                                                          []   
Make                                                                    \nAudi\n   
Model                                                               [\n, A1, \n]   
Offer Number                                                        [\nGZW086\n]   
First Registration                                                           NaN   
Body Color                                                       [\n, Black, \n]   
Paint Type                                                        [\nMetallic\n]   
Body Color Original                                          [\nMythosschwarz\n]   
Upholstery                                                    [\nCloth, Black\n]   
Body                                                           [\n, Compact, \n]   
Nr. of Doors                                                             [\n5\n]   
Nr. of Seats                                                             [\n5\n]   
Model Code                                                        [\n0588/BNN\n]   
Gearing Type                                                    [\n, Manual, \n]   
Displacement                                                        [\n999 cc\n]   
Cylinders                                                                    NaN   
Weight                                                            [\n1,165 kg\n]   
Drive chain                                                                  NaN   
Fuel                                                          [\n, Super 95, \n]   
Consumption                    [[4.6 l/100 km (comb)], [5.8 l/100 km (city)],...   
CO2 Emission                                           [\n104 g CO2/km (comb)\n]   
Emission Class                                                [\nEuro 6d-TEMP\n]   
\nComfort & Convenience\n      [Air conditioning, Armrest, Automatic climate ...   
\nEntertainment & Media\n      [Bluetooth, Hands-free equipment, MP3, On-boar...   
\nExtras\n                                         [Alloy wheels, Voice Control]   
\nSafety & Security\n          [ABS, Central door lock, Daytime running light...   
description                    [\n, Assistenzsysteme:,  Spracheingabesystem, ...   
Emission Label                                                   [\n4 (Green)\n]   
Gears                                                                        NaN   
Country version                            

In [71]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15919 entries, 0 to 15918
Data columns (total 54 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   url                            15919 non-null  object 
 1   make_model                     15919 non-null  object 
 2   short_description              15873 non-null  object 
 3   body_type                      15859 non-null  object 
 4   price                          15919 non-null  int64  
 5   vat                            11406 non-null  object 
 6   km                             15919 non-null  object 
 7   registration                   15919 non-null  object 
 8   prev_owner                     9091 non-null   object 
 9   kW                             0 non-null      float64
 10  hp                             15919 non-null  object 
 11  Type                           15917 non-null  object 
 12  Previous Owners                9279 non-null  

In [72]:
df1 = df.copy()

In [73]:
df1.drop('url', axis = 1, inplace = True)
df1.drop('make_model', axis = 1, inplace = True)
df1.drop('short_description', axis = 1, inplace = True)
df1.drop('kW', axis = 1, inplace = True)
df1.drop('registration', axis = 1, inplace = True)
df1.drop('Previous Owners', axis = 1, inplace = True)
df1.drop('Next Inspection', axis = 1, inplace = True)
df1.drop('Inspection new', axis = 1, inplace = True)
df1.drop('Full Service', axis = 1, inplace = True)
df1.drop('null', axis = 1, inplace = True)
df1.drop('Body Color Original', axis = 1, inplace = True)
df1.drop('Body', axis = 1, inplace = True)
df1.drop('Model Code', axis = 1, inplace = True)
df1.drop('Weight', axis = 1, inplace = True)
df1.drop('description', axis = 1, inplace = True)
df1.drop('Emission Label', axis = 1, inplace = True)
df1.drop('Country version', axis = 1, inplace = True)
df1.drop('Electricity consumption', axis = 1, inplace = True)
df1.drop('Last Service Date', axis = 1, inplace = True)
df1.drop('Other Fuel Types', axis = 1, inplace = True)
df1.drop('Availability', axis = 1, inplace = True)
df1.drop('Last Timing Belt Service Date', axis = 1, inplace = True)
df1.drop('Available from', axis = 1, inplace = True)
df1.drop('Non-smoking Vehicle', axis = 1, inplace = True)

In [74]:
df1.drop('Offer Number', axis = 1, inplace = True)

In [75]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15919 entries, 0 to 15918
Data columns (total 29 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   body_type                15859 non-null  object
 1   price                    15919 non-null  int64 
 2   vat                      11406 non-null  object
 3   km                       15919 non-null  object
 4   prev_owner               9091 non-null   object
 5   hp                       15919 non-null  object
 6   Type                     15917 non-null  object
 7   Warranty                 10499 non-null  object
 8   Make                     15919 non-null  object
 9   Model                    15919 non-null  object
 10  First Registration       14322 non-null  object
 11  Body Color               15322 non-null  object
 12  Paint Type               10147 non-null  object
 13  Upholstery               12199 non-null  object
 14  Nr. of Doors             15707 non-nul

In [76]:
df1.body_type.describe()

count      15859
unique         9
top       Sedans
freq        7903
Name: body_type, dtype: object

In [77]:
df1.body_type.unique()

array(['Sedans', 'Station wagon', 'Compact', 'Other', 'Coupe', 'Van',
       'Off-Road', 'Convertible', None, 'Transporter'], dtype=object)

In [78]:
df1.km

0        56,013 km
1        80,000 km
2        83,450 km
3        73,000 km
4        16,200 km
           ...    
15914         - km
15915     9,900 km
15916        15 km
15917        10 km
15918         - km
Name: km, Length: 15919, dtype: object

In [79]:
df3 = df.copy()

In [80]:
df3.km

0        56,013 km
1        80,000 km
2        83,450 km
3        73,000 km
4        16,200 km
           ...    
15914         - km
15915     9,900 km
15916        15 km
15917        10 km
15918         - km
Name: km, Length: 15919, dtype: object

In [81]:
df3.km = df3.km.astype('str').replace('- km', np.nan)

In [82]:
df3.km

0        56,013 km
1        80,000 km
2        83,450 km
3        73,000 km
4        16,200 km
           ...    
15914          NaN
15915     9,900 km
15916        15 km
15917        10 km
15918          NaN
Name: km, Length: 15919, dtype: object

In [83]:
df3.km  = df3.km.str.strip(" km")

In [84]:
df3.km = df3.km.str.replace(",", "")

In [85]:
df3.km

0        56013
1        80000
2        83450
3        73000
4        16200
         ...  
15914      NaN
15915     9900
15916       15
15917       10
15918      NaN
Name: km, Length: 15919, dtype: object

In [86]:
df3.km = df3.km.astype(float)

In [87]:
df3.km

0        56013.0
1        80000.0
2        83450.0
3        73000.0
4        16200.0
          ...   
15914        NaN
15915     9900.0
15916       15.0
15917       10.0
15918        NaN
Name: km, Length: 15919, dtype: float64

In [88]:
df1.km = df3.km

In [89]:
df[df.km == "- km"].sample(3).T

5683   \
url                            https://www.autoscout24.com//offers/audi-a3-sp...   
make_model                                                               Audi A3   
short_description                                        Sportback 35 TFSI 110kW   
body_type                                                                Compact   
price                                                                      25272   
vat                                                                         None   
km                                                                          - km   
registration                                                                 -/-   
prev_owner                                                                  None   
kW                                                                           NaN   
hp                                                                        110 kW   
Type                                                         [, New, , Gasoline]   
Previous Owners                                                              NaN   
Next Inspection                                                              NaN   
Inspection new                                                               NaN   
Warranty                                                                     NaN   
Full Service                                                                 NaN   
Non-smoking Vehicle                                                          NaN   
null                                                                          []   
Make                                                                    \nAudi\n   
Model                                                               [\n, A3, \n]   
Offer Number                                                       [\n2386559\n]   
First Registration                                                           NaN   
Body Color                                                                   NaN   
Paint Type                                                                   NaN   
Body Color Original                                                          NaN   
Upholstery                                                                   NaN   
Body                                                           [\n, Compact, \n]   
Nr. of Doors                                                             [\n5\n]   
Nr. of Seats                                                             [\n5\n]   
Model Code                                                                   NaN   
Gearing Type                                                    [\n, Manual, \n]   
Displacement                                                      [\n1,495 cc\n]   
Cylinders                                                                [\n4\n]   
Weight                                                            [\n1,230 kg\n]   
Drive chain                                                          [\nfront\n]   
Fuel                                                          [\n, Gasoline, \n]   
Consumption                    [[5 l/100 km (comb)], [6 l/100 km (city)], [4 ...   
CO2 Emission                                                                 NaN   
Emission Class                                                               NaN   
\nComfort & Convenience\n                                                    NaN   
\nEntertainment & Media\n                                                    NaN   
\nExtras\n                                                                   NaN   
\nSafety & Security\n                                                        NaN   
description                                                                   []   
Emission Label                                                               NaN   
Gears                                                                    [\n6\n]   
Country version                            

In [90]:
df3.kw

AttributeError: 'DataFrame' object has no attribute 'kw'

In [ ]:
df1.rename(columns={'hp':'k'}, inplace = True )

In [ ]:
df1.hp = df3.kW

In [ ]:
df3.kw

In [ ]:
df1.kW = df3.kw

In [ ]:
df1.kW

In [ ]:
df1.sample(3).T

In [ ]:
df1.kW = df1.kW.str.strip(' kW')

In [ ]:
df1.kW

In [ ]:
df1.kW.sample(5)

In [ ]:
df1.kW = df1.kW.replace(0,np.nan)

In [ ]:
df1.kW = df1.kW.astype(float)

In [ ]:
df1.kW

In [ ]:
df1.kW.isnull().value_counts()

In [ ]:
df1.prev_owner

In [ ]:
df1.prev_owner = df1.prev_owner.str.strip(' previous owner')

In [ ]:
df1.prev_owner

In [ ]:
df1.prev_owner = df1.prev_owner.astype(float)

In [ ]:
df1.prev_owner

In [ ]:
df1.info()

### Things Needs to Be Done

1. Think about removing url column.
2. **Get rid of the column "make_model"**
3. **short description
4. Body_type : There are 60 nulls values that we have to deal with.
5. Be aware of outliers such as min. in price column.
6. Vat column : Change it to deductible or not deductible
7. For the km column remove the "km" from the rows and change the data type to int or float.
8. **Drop the registration column because there is another registratio colum which only contains years.**
9. **From prev_owner column remove the part of "previous owner" and change data type to integer. There are 6828 null values.Remove the "Previous owners" column.**
10. **Drop the column "kW"**
11. Change the "hp" column name to kW.Change the data type.
12. Clean the "Type" column.
13. **Drop the columns about inspectation.**
14. Come back to the Warranty column
15. **Drop the Full service column.**
16. **Drop the Non_smoking column.**
17. **Drop the null column.**
18. Clean both make and model columns.
19. Come back to the offer number column again.
20. Change the First registration column name as Registration year
21. **Drop the Body color original**
22. Organize Upholstery column
23. **Drop the Body column.**
24. Check the number of doors
25. For the number od seats, clean '\n'
26. **Drop the model code number  column**
27. Clean the outliers for gearing type column
28. For the Cyclinder column clean the outliers.
29. **You might drop Weight column(!)**
30. Make the small groups for Fuel column.
31. We might need the 'Consumption' to be in 3 different column.
32. For the CO2 Emission, get rid of the unnecessary string parts and change the column name.
33. For the Emission Class column be aware of empty brackets which are null valus and get rid of unnecessary inf.
34. Come back to the \nComfort & Convenience\n column.
35. Come back to '\nEntertainment & Media\n' column.
36. Extras column*
37. Safety & Security column*
38. **Drop the description column.**
39. **Drop the emission column.**
40. Gears column*
41. **Drop rest of the columns.**

In [ ]:
df1.Type

In [ ]:
df3.Model

In [ ]:
df3.Model[:][1]

In [ ]:
df3.Make

In [ ]:
df2 = df1.copy()

In [ ]:
df2.Make = df2.Make.str.strip('\n')

In [ ]:
df1.Make = df2.Make

In [ ]:
df1.Make

In [ ]:
df3.Model

In [ ]:
#for i in df3.Model:
    #print(i[1])

In [ ]:
df2.Model.strip('\n')

In [ ]:
df3.Model

In [ ]:
df3.Model=df3.Model.str[1].str.strip('\n')

In [ ]:
df3.Model

In [ ]:
df1.Model = df3.Model

In [ ]:
df1.sample(3).T

In [ ]:
df4 = df.copy()

In [ ]:
df4.Model = df4.Model.str[1]

In [ ]:
df4.Model

In [ ]:
df1.Type

In [ ]:
df1.Type.describe()

In [ ]:
df1.Type.str[1].value_counts()

In [ ]:
df3.Type = df1.Type.str[1]

In [ ]:
df1.Type = df3.Type

In [ ]:
df3.Type.value_counts()

In [ ]:
df1.Warranty.describe()

In [ ]:
df3.Warranty

In [ ]:
df1.sample(3).T

In [ ]:
df3['Warranty'].str.contains("months").value_counts()

In [ ]:
df1.drop('Warranty', axis = 1, inplace = True)

In [ ]:
df1.info()

In [ ]:
df1.Type

In [ ]:
df1.rename(columns={'Type':'Condition'}, inplace = True)

In [ ]:
df1.rename(columns={'body_type':'Body_Type'}, inplace = True)
df1.rename(columns={'price':'Price'}, inplace = True)
df1.rename(columns={'vat':'Vat'}, inplace = True)
df1.rename(columns={'prev_owner':'Previous_Owner'}, inplace = True)
df1.rename(columns={'First Registration':'Registration_Year'}, inplace = True)
df1.rename(columns={'Body Color':'Body_Color'}, inplace = True)
df1.rename(columns={'Paint Type':'Paint_Type'}, inplace = True)
df1.rename(columns={'Nr. of Doors':'Nr_of_Doors'}, inplace = True)
df1.rename(columns={'Nr. of Seats':'Nr_of_Seats'}, inplace = True)
df1.rename(columns={'Gearing Type':'Gearing_Type'}, inplace = True)
df1.rename(columns={'CO2 Emission':'CO2_Emission'}, inplace = True)
df1.rename(columns={'Emission Class':'Emission_Class'}, inplace = True)
df1.rename(columns={'\nComfort & Convenience\n':'Comfort_Convenience'}, inplace = True)
df1.rename(columns={'\nEntertainment & Media\n':'Entertainment_Media'}, inplace = True)
df1.rename(columns={'\nExtras\n':'Extras'}, inplace = True)
df1.rename(columns={'\nSafety & Security\n':'Safety & Security'}, inplace = True)

In [ ]:
df1.rename(columns={'Safety & Security':'Safety_Security'}, inplace = True)

In [ ]:
df1.rename(columns={'Drive chain':'Drive_Chain'}, inplace = True)

In [ ]:
df1.rename(columns={'km':'Km'}, inplace = True)

In [ ]:
df1.sample(3).T

In [ ]:
df1.First_Registration = df3['First Registration'].str[1]

In [ ]:
df1.sample(3).T

1. Create age column.


### Things Needs to Be Done

1. Think about removing url column.
2. **Get rid of the column "make_model"**
3. **short description
4. Body_type : There are 60 nulls values that we have to deal with.
5. Be aware of outliers such as min. in price column.
6. Vat column : Change it to deductible or not deductible
7. For the km column remove the "km" from the rows and change the data type to int or float.
8. **Drop the registration column because there is another registratio colum which only contains years.**
9. **From prev_owner column remove the part of "previous owner" and change data type to integer. There are 6828 null values.Remove the "Previous owners" column.**
10. **Drop the column "kW"**
11. Change the "hp" column name to kW.Change the data type.
12. Clean the "Type" column.
13. **Drop the columns about inspectation.**
14. Come back to the Warranty column
15. **Drop the Full service column.**
16. **Drop the Non_smoking column.**
17. **Drop the null column.**
18. Clean both make and model columns.
19. Come back to the offer number column again.
20. Change the First registration column name as Registration year
21. **Drop the Body color original**
22. Organize Upholstery column
23. **Drop the Body column.**
24. Check the number of doors
25. For the number od seats, clean '\n'
26. **Drop the model code number  column**
27. Clean the outliers for gearing type column
28. For the Cyclinder column clean the outliers.
29. **You might drop Weight column(!)**
30. Make the small groups for Fuel column.
31. We might need the 'Consumption' to be in 3 different column.
32. For the CO2 Emission, get rid of the unnecessary string parts and change the column name.
33. For the Emission Class column be aware of empty brackets which are null valus and get rid of unnecessary inf.
34. Come back to the \nComfort & Convenience\n column.
35. Come back to '\nEntertainment & Media\n' column.
36. Extras column*
37. Safety & Security column*
38. **Drop the description column.**
39. **Drop the emission column.**
40. Gears column*
41. **Drop rest of the columns.**

In [ ]:
df1.Body_Color = df1.Body_Color.str[1]

In [ ]:
df1.Body_Color.describe()

In [ ]:
df1.Body_Color.isnull().sum()

In [ ]:
 a = df1.Upholstery

In [ ]:
for i in a :
    if type(i) == list:
        print(i[0])

In [ ]:
df1["Upholstery"] = df1["Upholstery"].apply(lambda item: item[0] if type(item) == list else item)

In [ ]:
df1.Upholstery

In [ ]:
df1.sample(3).T

In [ ]:
df1.Upholstery.str[1]

In [ ]:
df1.Upholstery = df1.Upholstery.str.strip("\n")

In [ ]:
df1.Upholstery.value_counts()

In [ ]:
df1.Paint_Type = df1["Paint_Type"].apply(lambda item: item[0] if type(item) == list else item)

In [ ]:
df1.Paint_Type = df1.Paint_Type.str.strip('\n')

In [ ]:
df1.Paint_Type

In [ ]:
df1.Paint_Type.describe()

In [ ]:
df1.Paint_Type.isnull().sum()

In [ ]:
df1.Paint_Type.value_counts()

In [ ]:
df1.Paint_Type

In [ ]:
df1.Nr_of_Doors = df1.Nr_of_Doors.apply(lambda item: item[0] if type(item) == list else item)

In [ ]:
df1.Nr_of_Doors = df1.Nr_of_Doors.str.strip('\n')

In [ ]:
df1.Nr_of_Seats = df1.Nr_of_Seats.apply(lambda item: item[0] if type(item) == list else item)

In [ ]:
df1.Nr_of_Seats = df1.Nr_of_Seats.str.strip('\n')

In [ ]:
df1.sample(3).T

In [ ]:
df1.Gearing_Type = df1.Gearing_Type.apply(lambda item: item[1] if type(item) == list else item)
df1.Gearing_Type = df1.Gearing_Type.str.strip('\n')

In [ ]:
df1.Gearing_Type = df3["Gearing Type"].str[1]

In [ ]:
df3['Gearing Type']

In [ ]:
df1.Gearing_Type = df1.Gearing_Type.str.strip('\n')

In [ ]:
df1.Gearing_Type

In [ ]:
df1.Displacement = df3.Displacement

In [ ]:
df1.Displacement.apply(lambda item: item[0] if type(item) == list else item)


In [ ]:
df1.Displacement = df1.Displacement.str[0].str.strip('\n').str.strip('cc')

In [ ]:
df1.Displacement.value_counts()

In [ ]:
df1.Displacement = df3.Displacement

In [ ]:
df1.Displacement  = df1.Displacement.str.replace(',','')

In [ ]:
df1.Displacement.astype(float)

In [ ]:
df1.Cylinders = df1.Cylinders.str[0].str.strip('\n').astype(float)

In [ ]:
df1.Cylinders.astype(float)

In [ ]:
df1.Drive_Chain = df1.Drive_Chain.str[0].str.strip('\n')

In [ ]:
df1.Fuel = df1.Fuel.str[1].str.strip('\n')

In [ ]:
df1.Fuel = df3.Fuel

In [ ]:
df1.Fuel.str[1].str.strip('\n')

In [ ]:
df1.Fuel.value_counts()

In [ ]:
df1.Fuel.describe()

In [ ]:
df1.Fuel.apply(lambda x : "Diesel" if 'Diesel' in x elif "Gasoline" in x "Gasoline" else "Gasoline")

In [ ]:
df1.Fuel

In [ ]:
df1.Fuel = df3.Fuel

In [ ]:
df3.Fuel = df1.Fuel

In [ ]:
df1.Fuel = df1.Fuel.apply(lambda x : "Diesel" if 'Diesel' in x else x)

In [ ]:
df1.Fuel.value_counts()

In [ ]:
df1.Fuel.describe()

In [ ]:
df1.Fuel = df1.Fuel.apply(lambda x : "Gasoline" if 'Benzine' in x else x)

In [ ]:
df1.Fuel = df1.Fuel.apply(lambda x : "Gasoline" if 'Super' in x else x)

In [ ]:
df1.Fuel = df1.Fuel.apply(lambda x : "LPG" if 'LPG' in x else "Others")

In [ ]:
df1.Fuel = df1.Fuel.apply(lambda x : "Gasoline" if "Benzine" in x else x)

In [ ]:
df1.Fuel

In [ ]:
df1.Fuel = df1.Fuel.apply(lambda x : "Diesel" if "Diesel" in x else x )

In [ ]:
df1.Fuel = df1.Fuel.apply(Fuel_type)

In [ ]:
def Fuel_type(x):
    if "Benzine" in x :
        return "Gasoline"
    elif "Gasoline" in x :
        return "Gasoline"
    elif "Super" in x :
        return "Gasoline"
    elif "Diesel" in x :
        return "Diesel"
    elif "LPG" in x :
        return "LPG"
    else :
        return "Others"

In [ ]:
df1.Fuel.value_counts()

In [ ]:
df1.Consumption

In [ ]:
df1['Combined'] = df1.Consumption.str[0].str[0]

In [ ]:
df1.Consumption.str[0].str.split().str[0]

In [ ]:
df1.Consumption.str[0].str[0].str.rstrip('l/100 km (comb)')

In [ ]:
df1.Consumption.iloc[4]

In [ ]:
df1.Consumption.str[0].str[0]

In [ ]:
df1.Combined

In [ ]:
df1[df1.Consumption.str[0].str[0].str.split().str[0].str.contains(".1")]

In [ ]:
df1.Combined = df1.Consumption.str[0].str[0].str.split().str[0]

In [ ]:
df1.Combined=df1.Combined.astype(float)

In [ ]:
df1.City_Cons = df1.Consumption.str[1].str[0].str.split().str[0].astype(float)

In [ ]:
df1.City_Cons

In [ ]:
df1.Highway_Cons = df1.Consumption.str[2].str[0].str.split().str[0].astype(float)

In [ ]:
df1.Highway_Cons

In [ ]:
df1.rename(columns = {"Combined" : "Combined_Cons"}, inplace = True)

In [ ]:
df1.drop('Consumption', axis = 1, inplace = True)

In [ ]:
df1.sample(3).T

In [ ]:
df1.CO2_Emission = df3['CO2 Emission']

In [ ]:
df1.CO2_Emission = df1.CO2_Emission.str[0]

In [ ]:
df1.CO2_Emission = df1.CO2_Emission.str.strip(' g CO2/km (comb)\n').str.replace(',',"")

In [ ]:
df1.CO2_Emission = df1.CO2_Emission.astype(float)

In [ ]:
df1.Emission_Class = df1.Emission_Class.str[0].str.strip('\n')

In [ ]:
df1.Emission_Class.value_counts()

In [ ]:
df1.columns

In [ ]:
df1.City_Cons

In [ ]:
df1.T

In [ ]:
df1.Highway_Cons

In [ ]:
df1.info()

In [91]:
df1

body_type  price               vat       km         prev_owner      hp  \
0        Sedans  15770    VAT deductible  56013.0  2 previous owners   66 kW   
1        Sedans  14500  Price negotiable  80000.0               None  141 kW   
2        Sedans  14640    VAT deductible  83450.0   1 previous owner   85 kW   
3        Sedans  14500              None  73000.0   1 previous owner   66 kW   
4        Sedans  16790              None  16200.0   1 previous owner   66 kW   
...         ...    ...               ...      ...                ...     ...   
15914       Van  39950    VAT deductible      NaN               None  147 kW   
15915       Van  39885    VAT deductible   9900.0   1 previous owner  165 kW   
15916       Van  39875    VAT deductible     15.0   1 previous owner  146 kW   
15917       Van  39700    VAT deductible     10.0               None  147 kW   
15918       Van  40999    VAT deductible      NaN   1 previous owner  165 kW   

                                                    Type  \
0                [, Used, , Diesel (Particulate Filter)]   
1                                   [, Used, , Gasoline]   
2                [, Used, , Diesel (Particulate Filter)]   
3                [, Used, , Diesel (Particulate Filter)]   
4                [, Used, , Diesel (Particulate Filter)]   
...                                                  ...   
15914             [, New, , Diesel (Particulate Filter)]   
15915  [, Used, , Super 95 / Super Plus 98 (Particula...   
15916                       [, Pre-registered, , Diesel]   
15917                       [, Pre-registered, , Diesel]   
15918                      [, Demonstration, , Super 95]   

                               Warranty         Make             Model  ...  \
0               [\n, \n, \n4 (Green)\n]     \nAudi\n      [\n, A1, \n]  ...   
1                                   NaN     \nAudi\n      [\n, A1, \n]  ...   
2      [\n, \n, \n99 g CO2/km (comb)\n]     \nAudi\n      [\n, A1, \n]  ...   
3                                   NaN     \nAudi\n      [\n, A1, \n]  ...   
4                  [\n, \n, \nEuro 6\n]     \nAudi\n      [\n, A1, \n]  ...   
...                                 ...          ...               ...  ...   
15914                     \n24 months\n  \nRenault\n  [\n, Espace, \n]  ...   
15915              [\n, \n, \nEuro 6\n]  \nRenault\n  [\n, Espace, \n]  ...   
15916           [\n, \n, \n4 (Green)\n]  \nRenault\n  [\n, Espace, \n]  ...   
15917                                \n  \nRenault\n  [\n, Espace, \n]  ...   
15918                               NaN  \nRenault\n  [\n, Espace, \n]  ...   

       Drive chain                                               Fuel  \
0      [\nfront\n]              [\n, Diesel (Particulate Filter), \n]   
1      [\nfront\n]                                 [\n, Gasoline, \n]   
2      [\nfront\n]              [\n, Diesel (Particulate Filter), \n]   
3              NaN              [\n, Diesel (Particulate Filter), \n]   
4      [\nfront\n]              [\n, Diesel (Particulate Filter), \n]   
...            ...                                                ...   
15914  [\nfront\n]              [\n, Diesel (Particulate Filter), \n]   
15915  [\nfront\n]  [\n, Super 95 / Super Plus 98 (Particulate Fil...   
15916  [\nfront\n]                                   [\n, Diesel, \n]   
15917  [\nfront\n]                                   [\n, Diesel, \n]   
15918    [\n4WD\n]                                 [\n, Super 95, \n]   

                                             Consumption  \
0      [[3.8 l/100 km (comb)], [4.3 l/100 km (city)],...   
1      [[5.6 l/100 km (comb)], [7.1 l/100 km (city)],...   
2      [[3.8 l/100 km (comb)], [4.4 l/100 km (city)],...   
3      [[3.8 l/100 km (comb)], [4.3 l/100 km (city)],...   
4      [[4.1 l/100 km (comb)], [4.6 l/100 km (city)],...   
...                                                  ...   
15914  [[5.3 l/100 km (comb)], [6.2 l/100 km (city)],...   
15915  [\n, 7.4 l/